File: src/examples/simple_sdp_cone/simple_sdp_cone.cpp
// Optimize a simple problem with an optimal solution of (0.5,.25)

#include <iostream>
#include <iomanip>
#include "optizelle/optizelle.h"
#include "optizelle/vspaces.h"
#include "optizelle/json.h"

// Create some type shortcuts
using Optizelle::Rm;
using Optizelle::SQL;
typedef double Real;

// Define a simple objective where
//
// f(x,y)=-x+y
//
struct MyObj : public Optizelle::ScalarValuedFunction <Real,Rm> {
    typedef Rm <Real> X;

    // Evaluation
    double eval(X::Vector const & x) const {
        return -x[0]+x[1];
    }

    // Gradient
    void grad(
        X::Vector const & x,
        X::Vector & grad
    ) const {
        grad[0]=Real(-1.);
        grad[1]=Real(1.);
    }

    // Hessian-vector product
    void hessvec(
        X::Vector const & x,
        X::Vector const & dx,
        X::Vector & H_dx
    ) const {
        H_dx[0]= Real(0.);
        H_dx[1]= Real(0.);
    }
};

// Define a simple SDP inequality
//
// h(x,y) = [ y x ] >= 0
//          [ x 1 ]
//
struct MyIneq : public Optizelle::VectorValuedFunction <Real,Rm,SQL> {
    typedef Rm <Real> X;
    typedef SQL <Real> Z;

    // z=h(x)
    void eval(
        X::Vector const & x,
        Z::Vector & z
    ) const {
        z(1,1,1)=x[1];
        z(1,1,2)=x[0];
        z(1,2,1)=x[0];
        z(1,2,2)=Real(1.);
    }

    // z=h'(x)dx
    void p(
        X::Vector const & x,
        X::Vector const & dx,
        Z::Vector & z
    ) const {
        z(1,1,1)=dx[1];
        z(1,1,2)=dx[0];
        z(1,2,1)=dx[0];
        z(1,2,2)=Real(0.);
    }

    // xhat=h'(x)*dz
    void ps(
        X::Vector const & x,
        Z::Vector const & dz,
        X::Vector & xhat
    ) const {
        xhat[0]= dz(1,1,2)+dz(1,2,1);
        xhat[1]= dz(1,1,1);
    }

    // xhat=(h''(x)dx)*dz
    void pps(
        X::Vector const & x,
        X::Vector const & dx,
        Z::Vector const & dz,
        X::Vector & xhat
    ) const {
        X::zero(xhat);
    }
};

int main(int argc,char* argv[]){
    // Create some type shortcuts
    typedef Rm <Real>::Vector Rm_Vector;
    typedef SQL <Real>::Vector SQL_Vector;

    // Read in the name for the input file
    if(argc!=2) {
        std::cerr << "simple_sdp_cone <parameters>" << std::endl;
        exit(EXIT_FAILURE);
    }
    auto fname = argv[1];

    // Generate an initial guess for the primal
    auto x = Rm_Vector({1.2,3.1});

    // Allocate memory for the dual
    auto z = SQL_Vector ({Optizelle::Cone::Semidefinite},{2});

    // Create an optimization state
    Optizelle::InequalityConstrained <Real,Rm,SQL>::State::t state(x,z);

    // Read the parameters from file
    Optizelle::json::InequalityConstrained <Real,Rm,SQL>::read(fname,state);

    // Create a bundle of functions
    Optizelle::InequalityConstrained <Real,Rm,SQL>::Functions::t fns;
    fns.f.reset(new MyObj);
    fns.h.reset(new MyIneq);

    // Solve the optimization problem
    Optizelle::InequalityConstrained <Real,Rm,SQL>
        ::Algorithms::getMin(Optizelle::Messaging::stdout,fns,state);

    // Print out the reason for convergence
    std::cout << "The algorithm converged due to: " <<
        Optizelle::OptimizationStop::to_string(state.opt_stop) << std::endl;

    // Print out the final answer
    std::cout << std::setprecision(16) << std::scientific
        << "The optimal point is: (" << state.x[0] << ','
        << state.x[1] << ')' << std::endl;

    // Write out the final answer to file
    Optizelle::json::InequalityConstrained <Real,Rm,SQL>
        ::write_restart("solution.json",state);

    // Successful termination
    return EXIT_SUCCESS;
}

File: src/examples/simple_sdp_cone/simple_sdp_cone.m
% Optimize a simple problem with an optimal solution of (0.5,.25)

function simple_sdp_cone(fname)
    % Read in the name for the input file
    if nargin ~=1
        error('simple_sdp_cone <parameters>');
    end

    % Execute the optimization
    main(fname);
end

% Define a simple objective where 
% 
% f(x,y)=-x+y
%
function self = MyObj()

    % Evaluation 
    self.eval = @(x) -x(1)+x(2);

    % Gradient
    self.grad = @(x) [
        -1.;
        1.];

    % Hessian-vector product
    self.hessvec = @(x,dx) [
        0;
        0];
end

% Define a simple SDP inequality 
%
% h(x,y) = [ y x ] >= 0
%          [ x 1 ]
%
function self = MyIneq()

    % y=h(x) 
    self.eval = @(x)MyIneq_eval(x);

    % y=h'(x)dx
    self.p = @(x,dx)MyIneq_p(x,dx);

    % z=h'(x)*dy
    self.ps = @(x,dy) [
        dy.data{1}(2,1)+dy.data{1}(1,2);
        dy.data{1}(1,1)];

    % z=(h''(x)dx)*dy
    self.pps = @(x,dx,dy) [
        0;
        0];
end

% y=h(x) 
function y=MyIneq_eval(x)
    global Optizelle;
    y = Optizelle.SQL.create([Optizelle.Cone.Semidefinite],[2]);
    y.data{1} = [
        x(2) x(1);
        x(1) 1.  ];
end

% y=h'(x)dx
function y=MyIneq_p(x,dx)
    global Optizelle;
    y = Optizelle.SQL.create([Optizelle.Cone.Semidefinite],[2]);
    y.data{1} = [
        dx(2) dx(1);
        dx(1) 0.   ];
end

% Actually runs the program
function main(fname)

    % Grab the Optizelle library
    global Optizelle;
    setupOptizelle();

    % Generate an initial guess for the primal
    x = [1.2; 3.1];

    % Generate an initial guess for the dual
    z = Optizelle.SQL.create([Optizelle.Cone.Semidefinite],[2]);

    % Create an optimization state
    state=Optizelle.InequalityConstrained.State.t( ...
        Optizelle.Rm,Optizelle.SQL,Optizelle.Messaging,x,z);

    % Read the parameters from file
    state=Optizelle.json.InequalityConstrained.read( ...
        Optizelle.Rm,Optizelle.SQL,Optizelle.Messaging,fname,state);
    
    % Create a bundle of functions
    fns=Optizelle.InequalityConstrained.Functions.t;
    fns.f=MyObj();
    fns.h=MyIneq();

    % Solve the optimization problem
    state=Optizelle.InequalityConstrained.Algorithms.getMin( ...
        Optizelle.Rm,Optizelle.SQL,Optizelle.Messaging,fns,state);

    % Print out the reason for convergence
    fprintf('The algorithm converged due to: %s\n', ...
        Optizelle.StoppingCondition.to_string(state.opt_stop));

    % Print out the final answer
    fprintf('The optimal point is: (%e,%e)\n',state.x(1),state.x(2));

    % Write out the final answer to file
    Optizelle.json.InequalityConstrained.write_restart( ...
        Optizelle.Rm,Optizelle.SQL,Optizelle.Messaging,'solution.json',state);
end
